### PR TITLE
WIP: use associated type for WIDTH

### PR DESCRIPTION
**Pull request by almindor**
First attempted WIP for associated type usage to allow easier "any Sized + Copy" for data format.

NOTE: tested with `ST7789` and `display-interface-spi` using `u8` "hardcoded", didn't try to use `u16` anywhere yet.

### CODE REVIEW

**Comment by almindor** [comment]
NOTE: do not merge yet this is a WIP

**Comment by almindor** [comment]
Here's a "problem overview" of this approach: [rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=243499c8466184b764b466536513febe)

See comments, basically this design forces driver writers into having most of their code in the "width specific" implementation blocks, and only basic data construction that's internal can be put outside into the "generic width" driver parts.

I still haven't figured out a better way tho...

**Comment by almindor** [comment]
Bumped with a `DisplayFormat<'a>` enum that lists each possible type and uses `core::slice::Iter` for internal representation.

There's a runtime cost to this approach both in the `as_slice` conversions (which probably can be removed if we optimize inside the DI implementations) as well as the match checks tho. It does remove the "duplication" in the driver tho, see `assoc` branch on ST7789 [here](https://github.com/almindor/st7789/pull/7)

File: src/lib.rs [hunk 1]
```diff
@@ -21,13 +23,33 @@ pub enum DisplayError {
     CSError,
 }
 
+/// DI specific data format wrapper around slices of various widths
+/// Display drivers need to implement non-trivial conversions (e.g. with padding)
+/// as the hardware requires.
+pub enum DataFormat<'a> {
+    U8(&'a [u8]),
```
[the remark this answers is not in the document]
Comment (therealprof) [reply src/lib.rs:30]: I'm okay with slices for now. We can later always revisit this and add an iterator interface.

**Comment by therealprof** [comment]
Looks great to me on first glance. Let me try to convert a few drivers to it to check practicability.

**Comment by almindor** [comment]
For the iterator version I got blocked on https://github.com/rust-embedded/embedded-hal/issues/208

**Pull request merged**